### PR TITLE
[SERF-1677] Add recovery middleware for GRPC and HTTP server

### PR DESCRIPTION
**Pull request by laynax**
## Description

In this PR, we add a new HTTP middleware and a GRPC interceptor which both serves one purpose, panic recovery in the service endpoint level.

Based on this change, after a panic in an endpoint, the services will:
- Send a sentry event
- Log the panic error in `Fatal` level
- Exit the program

## Testing considerations

This was tested on [go-chassis dev](https://github.com/scribd/go-chassis/actions/runs/9567491587) env [applying both middlewares](https://github.com/scribd/go-chassis/commit/c19f3597ab4920678dfa88129b9859f0e432add8).
Sentry events (with both HTTP and GRPC panics) [here](https://scribd.sentry.io/issues/?project=1800793&referrer=sidebar&statsPeriod=7d) and datadog logs [here](https://app.datadoghq.com/logs?query=service%3Ago-chassis%20env%3Adevelopment%20-heartbeat%20status%3Ainfo%20level%3Dfatal%20&agg_m=count&agg_m_source=base&agg_t=count&cols=host%2Cservice&fromUser=true&messageDisplay=inline&refresh_mode=sliding&storage=hot&stream_sort=desc&viz=stream&from_ts=1718623084087&to_ts=1718795884087&live=true)

## Checklist

- [x] Prefixed the PR title with the JIRA ticket code 
- [x] Performed simple, atomic commits with [good commit messages][commit messages]
- [x] Verified that the commit history is linear and commits are squashed as necessary
- [x] Thoroughly tested the changes in `development` and/or `staging`
- [ ] Updated the `README.md` as necessary

## Related links

* https://pkg.go.dev/github.com/grpc-ecosystem/go-grpc-middleware/recovery
* https://github.com/grpc-ecosystem/go-grpc-middleware
* https://docs.sentry.io/platforms/go/usage/panics/
* [SERF-1677](https://scribdjira.atlassian.net/browse/SERF-1677)
* https://github.com/scribd/go-chassis/pull/166

[SERF-1677]: https://scribdjira.atlassian.net/browse/SERF-1677?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: pkg/interceptors/recovery.go [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+var recoveryOption = []grpcrecovery.Option{
+	grpcrecovery.WithRecoveryHandlerContext(func(ctx context.Context, rec interface{}) (err error) {
+		sentry.CurrentHub().Recover(rec)
+		sentry.Flush(time.Second * 5)
+
+		l, err := sdkloggercontext.Extract(ctx)
+		if err != nil {
+			debug.PrintStack()
+			stdlog.Printf("logger not found in context: %v\n", err)
+			stdlog.Fatalf("panic error: %v", rec)
```
Comment (fotos) [inline pkg/interceptors/recovery.go:41]: nit: to match the HTTP handler

```suggestion
			stdlog.Fatalf("grpc: panic error: %v", rec)
```

File: pkg/interceptors/recovery.go [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+// sends a sentry event, log in fatal level and halts the service.
+// IMPORTANT: This interceptor should be the last one in the interceptor chain.
+func RecoveryStreamServerInterceptor() grpc.StreamServerInterceptor {
+	return grpcrecovery.StreamServerInterceptor(recoveryOption...)
+}
+
+var recoveryOption = []grpcrecovery.Option{
+	grpcrecovery.WithRecoveryHandlerContext(func(ctx context.Context, rec interface{}) (err error) {
+		sentry.CurrentHub().Recover(rec)
+		sentry.Flush(time.Second * 5)
```
Comment (fotos) [inline pkg/interceptors/recovery.go:35]: Shouldn't this be wrapped in:

```
if rec := recover(); rec != nil {
}
```

I'm slightly confused as to how this is supposed to work for gRPC. 🤔

**Review by fotos — COMMENTED** [review]
Left a few comments/questions. 💬

File: pkg/interceptors/recovery.go (hunk 2 above)
Comment (laynax) [reply pkg/interceptors/recovery.go:35]: It is implemented in the [go-grpc-middleware](https://github.com/grpc-ecosystem/go-grpc-middleware/blob/62b7de50cda5a5d633f1013bfbe50e0f38db34ef/interceptors/recovery/interceptors.go#L25-L36) library when passing the option with `grpcrecovery.Option`.

File: pkg/interceptors/recovery.go (hunk 1 above)
Comment (laynax) [reply pkg/interceptors/recovery.go]: Done.

File: pkg/interceptors/recovery.go (hunk 2 above)
Comment (fotos) [reply pkg/interceptors/recovery.go:35]: Got it – thanks for explaining.

**Review by fotos — APPROVED** [review]
:+1: LGTM 🚀 

Tagging @Neurostep to re-review after PTO.

**Pull request merged**